bin/olly.ml

```diff
@@ -171,30 +190,38 @@ let trace format trace_filename exec_args =
       in
       let init () = () in
       let cleanup () = Trace.close trace_file in
-      olly ~runtime_begin ~runtime_end ~init ~cleanup exec_args
+      olly ~runtime_begin ~runtime_end ~init ~lifecycle ~cleanup exec_args
 
-let latency json output exec_args =
+let gc_stats json output exec_args =
   let current_event = Hashtbl.create 13 in
   let hist =
     H.init ~lowest_discernible_value:10 ~highest_trackable_value:10_000_000_000
       ~significant_figures:3
   in
+  let is_gc_phase phase =
+    match phase with
+    | "major" | "stw_leader" | "interrupt_remote" -> true
+    | _ -> false
+  in
   let runtime_begin ring_id ts phase =
     match Hashtbl.find_opt current_event ring_id with
     | None -> Hashtbl.add current_event ring_id (phase, Ts.to_int64 ts)
     | _ -> ()
   in
   let runtime_end ring_id ts phase =
     match Hashtbl.find_opt current_event ring_id with
-    | Some (saved_phase, saved_ts) when saved_phase = phase ->
+    | Some (saved_phase, saved_ts) when saved_phase = phase -> begin
         Hashtbl.remove current_event ring_id;
         let latency = Int64.to_int (Int64.sub (Ts.to_int64 ts) saved_ts) in
-        assert (H.record_value hist latency)
+        assert (H.record_value hist latency);
+        if is_gc_phase (Runtime_events.runtime_phase_name phase) then
```

Comment:
This logic doesn't look right to me. At `runtime_begin` you add any event to the hash table, not just the gc events. But the only event removed is a gc event. So if you ever have custom events, then according to this logic, the span for that event will never end. 

What you need is to only add events into the hash table at `runtime_begin` for gc events (using `is_gc_phase`). Given that the GC events are not nested, no need to check `is_gc_phase` here; the `when` clause is sufficient.

Reply:
Right, this makes a lot of sense. After going through the traces of more parallel programs, it seems olly was also tracking `domain_condition_wait` for latencies. This won't be the case if we do the check in the start of the event. Updated the PR.